### PR TITLE
cache range in PartitionKeyRange to avoid Range parsing json on every query

### PR DESCRIPTION
**Pull request by moderakh**
cache Range in PartitionKeyRange to avoid json parsing range on each query

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyRange.java [hunk 1]
```diff
@@ -76,20 +78,28 @@ public String getMinInclusive() {
...
-    public void setMaxExclusive(String maxExclusive) {
+    private void setMaxExclusive(String maxExclusive) {
         BridgeInternal.setProperty(this, "maxExclusive", maxExclusive);
     }
 
     public Range<String> toRange() {
-        return new Range<String>(this.getMinInclusive(), this.getMaxExclusive(), true, false);
+        if (this.range == null) {
+            synchronized (this) {
+                if (range != null) {
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyRange.java]: Should'n it be range  == null

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyRange.java [hunk 2]
```diff
@@ -76,20 +78,28 @@ public String getMinInclusive() {
...
 
-    public void setMaxExclusive(String maxExclusive) {
+    private void setMaxExclusive(String maxExclusive) {
         BridgeInternal.setProperty(this, "maxExclusive", maxExclusive);
     }
 
     public Range<String> toRange() {
-        return new Range<String>(this.getMinInclusive(), this.getMaxExclusive(), true, false);
+        if (this.range == null) {
+            synchronized (this) {
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyRange.java]: Loud thinking  - If toRange () method return same value for a PartitionKeyRange object , I am seeing a trade off between locking (making sure only one thread create object ) vs time saving if we dont do synchronized and letting two thread create object if they simultaneously comes in this block

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyRange.java (hunk 1 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyRange.java]: this was on earlier commit. already fixed.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyRange.java (hunk 2 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyRange.java]: as per Happen-Before relationship in java. To allow one thread sees the effect of the other, either this field has to be volatile or we should do synchronized lock.

as synchrnozied is only used on set assignment it doesn't have any cost on the get, however there is always a cost associated with get in volatile. synchronized doesn't add any cost on the get which is the intention.

Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyRange.java]: As i said in this particular scenario as both thread end up creating same value we can deviate from Happens-Before , how ever as you mentioned synchronized  doesn't add any cost then this is good . All set

**Pull request closed without merging**